File: Java/1110.delete-nodes-and-return-forest.java
/*
 * @lc app=leetcode id=1110 lang=java
 *
 * [1110] Delete Nodes And Return Forest
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    /*
    利用set存储需要删除的数字  
    root进行递归：如果需要删除，当前节点返回null, 并标记isRoot 和 deleted, 
    如果deleted为true, 则left & right child标记为isRoot = true
    time: O(n)
    space: O(n)
    */
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        List<TreeNode> res = new ArrayList<>();
        Set<Integer> set = new HashSet<>();
        for (int i : to_delete) set.add(i);
        helper(root, set, res, true);
        return res;
    }

    private TreeNode helper(TreeNode root, Set<Integer> set, List<TreeNode> res, boolean isRoot) {
        if (root == null) return null;
        boolean deleted = set.contains(root.val);
        if (isRoot && !deleted) res.add(root);
        root.left = helper(root.left, set, res, deleted);
        root.right = helper(root.right, set, res, deleted);
        return deleted ? null : root;
    }
}
// @lc code=end

File: Cpp/1110.delete-nodes-and-return-forest.cpp
/*
 * @lc app=leetcode id=1110 lang=cpp
 *
 * [1110] Delete Nodes And Return Forest
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        vector<TreeNode*> res;
        unordered_set<int> set(to_delete.begin(), to_delete.end());
        bool isRoot = (set.find(root->val) == set.end());
        helper(res, root, set, isRoot);
        return res;
    }
private:
    TreeNode* helper(vector<TreeNode*>& res, TreeNode* root, unordered_set<int>& set, bool& isRoot) {
        if (root == NULL) return NULL;
        bool deleted = set.find(root->val) != set.end();
        if(isRoot && !deleted) res.push_back(root);
        root->left = helper(res, root->left, set, deleted);
        root->right = helper(res, root->right, set, deleted);
        return deleted ? NULL : root;
    }
};
// @lc code=end

